File: skills/delegate/scripts/delegate_core/runs.py
"""Run records: where they live, their state, the machine-wide pool, removal."""

import json
import os
import shutil
import subprocess
import time
from datetime import datetime, timezone
from pathlib import Path

from .common import ACTIVE, DEFAULT_LIMITS, STARTING_GRACE, die, read_json, setting
from .worktree import remove_worktree


def runs_root():
    if setting("RUNS"):
        return Path(setting("RUNS"))
    try:
        top = subprocess.run(["git", "rev-parse", "--show-toplevel"], capture_output=True, text=True,
                             check=True).stdout.strip()
    except (OSError, subprocess.CalledProcessError):
        top = os.getcwd()
    return Path(top) / ".local/run/pi"



def all_runs():
    root = runs_root()
    if not root.is_dir():
        return []
    runs = [d for d in root.iterdir() if (d / "meta.json").is_file()]
    return sorted(runs, key=lambda d: (read_json(d / "meta.json", {}) or {}).get("startedNs", 0))



def resolve_run(ref):
    path = Path(ref)
    if (path / "meta.json").is_file():
        return path.resolve()
    root = runs_root()
    runs = all_runs()
    if ref == "last":
        if not runs:
            die(f"no runs under {root}")
        return runs[-1]
    if (root / ref / "meta.json").is_file():
        return root / ref
    matches = [d for d in runs if ref in d.name]
    if len(matches) != 1:
        die(f"run '{ref}' matched {len(matches)} runs under {root} (runs live under the git root of the "
            "directory where start ran; pass a run directory or set DELEGATE_RUNS)")
    return matches[0]



def pid_alive(pid):
    try:
        os.kill(pid, 0)
    except (ProcessLookupError, PermissionError, TypeError):
        return False
    return True



def supervisor_alive(run):
    try:
        pid = int((run / "pid").read_text())
        return pid_alive(pid) and b"_supervise" in Path(f"/proc/{pid}/cmdline").read_bytes()
    except (OSError, ValueError):
        return False



def agent_alive(run):
    """The agent's process group still exists, e.g. after its supervisor was killed."""
    try:
        pid = int((run / "agent.pid").read_text())
        return pid_alive(pid) and os.getpgid(pid) == pid
    except (OSError, ValueError):
        return False


def run_state(run):
    if (run / "exit_code").is_file():
        return (read_json(run / "summary.json", {}) or {}).get("state", "crashed")
    if supervisor_alive(run):
        return "running"
    if (run / "pid").is_file():
        return "crashed"
    # A supervisor that never started must not block writers in this workdir forever.
    started = (read_json(run / "meta.json", {}) or {}).get("startedEpoch", 0)
    return "starting" if time.time() - started <= STARTING_GRACE else "crashed"



def events(run):
    out = []
    try:
        with open(run / "events.jsonl", encoding="utf-8") as stream:
            for line in stream:
                try:
                    out.append(json.loads(line))
                except ValueError:
                    continue
    except OSError:
        pass
    return out



def status(run):
    meta = read_json(run / "meta.json", {}) or {}
    state = run_state(run)
    summary = read_json(run / "summary.json", {}) or {}
    out = {"run": meta.get("run", run.name), "name": meta.get("name"), "state": state,
           "agent": meta.get("agent", "pi"), "mode": meta.get("mode")}
    for key in ("parent", "worktree"):
        if meta.get(key):
            out[key] = meta[key]["path"] if key == "worktree" else meta[key]
    if summary:
        for key in ("elapsedSeconds", "attempts", "model", "turns", "files", "changes", "accept", "readOnlyViolation",
                    "tokens", "warning", "error"):
            if summary.get(key) not in (None, [], {}):
                out[key] = summary[key]
    else:
        evs = events(run)
        turns = [e for e in evs if e.get("e") == "turn"]
        out["elapsedSeconds"] = int(time.time() - meta.get("startedEpoch", time.time()))
        out["turns"] = len(turns)
        files = sorted({e["path"] for e in evs if e.get("e") in ("edit", "write") and e.get("path")})
        if files:
            out["files"] = files
        acts = [e for e in evs if e.get("e") not in ("turn", "result", "settled", "agent_end", "bash_done")]
        if state == "running" and acts:
            last = acts[-1]
            detail = last.get("cmd") or last.get("path") or last.get("arg") or last.get("detail") or ""
            out["last"] = last["e"] + (f": {detail}" if detail else "")
            try:
                at = datetime.strptime(last["at"], "%Y-%m-%dT%H:%M:%SZ").replace(tzinfo=timezone.utc)
                out["idleSeconds"] = int(time.time() - at.timestamp())
            except (KeyError, ValueError):
                pass
    result = run / "result.md"
    if result.is_file() and result.stat().st_size:
        out["result"] = str(result)
        out["resultChars"] = len(result.read_text(encoding="utf-8"))
    out["dir"] = str(run)
    return out



def unmerged_worktree(run):
    tree = (read_json(run / "meta.json", {}) or {}).get("worktree") or {}
    return bool(tree.get("path")) and Path(tree["path"]).exists() and not (run / ".applied").exists()



def remove_run(run):
    """Delete a run directory; its worktree goes with the last run that uses it."""
    meta = read_json(run / "meta.json", {}) or {}
    shutil.rmtree(run, ignore_errors=True)
    path = (meta.get("worktree") or {}).get("path")
    if path and not any(((read_json(r / "meta.json", {}) or {}).get("worktree") or {}).get("path") == path
                        for r in all_runs()):
        remove_worktree(meta)



def prune_expired():
    days = setting("KEEP_DAYS", "7")
    if not days.isdigit() or int(days) == 0:
        return
    cutoff = time.time() - int(days) * 86400
    for run in all_runs():
        done = run / "exit_code"
        if (run / ".delivered").is_file() and done.is_file() and done.stat().st_mtime < cutoff \
                and not unmerged_worktree(run):
            remove_run(run)



def state_dir():
    # Deliberately not a DELEGATE_* setting: a delegated agent must not opt out of the machine's pool.
    return Path(os.environ.get("XDG_STATE_HOME") or Path.home() / ".local/state") / "delegate"



def limit(name):
    value = setting(name, str(DEFAULT_LIMITS[name]))
    if not value.isdigit():
        die(f"DELEGATE_{name} must be a non-negative integer (0 = unlimited), got {value!r}")
    return int(value)



def machine_runs(slots):
    """Active runs on this machine, across projects; forget slots whose run has ended or vanished."""
    active = []
    for slot in slots.glob("*.slot"):
        try:
            run = Path(slot.read_text().strip())
        except OSError:
            continue
        if (run / "meta.json").is_file() and (run_state(run) in ACTIVE or agent_alive(run)):
            active.append(run)
        else:
            slot.unlink(missing_ok=True)
    return active



def capacity_error(agent, active):
    agents = [(read_json(run / "meta.json", {}) or {}).get("agent", "pi") for run in active]
    total, codex = limit("MAX_ACTIVE"), limit("MAX_CODEX")
    if total and len(active) >= total:
        reason = f"{len(active)} runs are active on this machine (DELEGATE_MAX_ACTIVE={total})"
    elif agent == "codex" and codex and agents.count("codex") >= codex:
        reason = f"{agents.count('codex')} Codex runs are active on this machine (DELEGATE_MAX_CODEX={codex})"
    else:
        return None
    listing = "".join(f"\n  {status(run)['elapsedSeconds']:>5}s {kind:<5} {run}" for run, kind in zip(active, agents))
    return (f"refusing to start: {reason}; collect results with wait before starting more, or stop runs "
            f"no longer needed{listing}")



def latest_in_chain(run):
    """A conversation is named by its first run; replies and apply act on its latest one."""
    runs = all_runs()
    while True:
        replies = [r for r in runs if (read_json(r / "meta.json", {}) or {}).get("parent") == run.name]
        if not replies:
            return run
        run = replies[-1]

File: crates/delegate/src/runs.rs
use crate::common::*;
use crate::worktree;
use serde_json::{json, Value};
use std::fs;
use std::path::{Path, PathBuf};

pub fn all_runs() -> Vec<PathBuf> {
    let Ok(entries) = fs::read_dir(runs_root()) else {
        return vec![];
    };
    let mut v = entries
        .flatten()
        .map(|e| e.path())
        .filter(|p| p.join("meta.json").is_file())
        .collect::<Vec<_>>();
    v.sort_by_key(|p| json(p.join("meta.json"))["startedNs"].as_u64().unwrap_or(0));
    v
}
pub fn resolve(reference: &str) -> Res<PathBuf> {
    let path = Path::new(reference);
    if path.join("meta.json").is_file() {
        return path.canonicalize().map_err(|e| e.to_string());
    }
    let root = runs_root();
    let runs = all_runs();
    if reference == "last" {
        return runs
            .last()
            .cloned()
            .ok_or_else(|| format!("no runs under {}", root.display()));
    }
    if root.join(reference).join("meta.json").is_file() {
        return Ok(root.join(reference));
    }
    let matches = runs
        .iter()
        .filter(|p| {
            p.file_name()
                .unwrap_or_default()
                .to_string_lossy()
                .contains(reference)
        })
        .collect::<Vec<_>>();
    if matches.len() != 1 {
        return Err(format!("run '{reference}' matched {} runs under {} (runs live under the git root of the directory where start ran; pass a run directory or set DELEGATE_RUNS)",matches.len(),root.display()));
    }
    Ok(matches[0].clone())
}
pub fn supervisor_alive(run: &Path) -> bool {
    let pid = read(run.join("pid")).trim().parse::<i32>().unwrap_or(0);
    pid_alive(pid)
        && fs::read(format!("/proc/{pid}/cmdline"))
            .is_ok_and(|x| x.windows(10).any(|w| w == b"_supervise"))
}
pub fn agent_alive(run: &Path) -> bool {
    let pid = read(run.join("agent.pid"))
        .trim()
        .parse::<i32>()
        .unwrap_or(0);
    pid_alive(pid) && unsafe { libc::getpgid(pid) == pid }
}
pub fn state(run: &Path) -> String {
    if run.join("exit_code").is_file() {
        let sum = json(run.join("summary.json"));
        return sum
            .get("state")
            .and_then(Value::as_str)
            .unwrap_or("crashed")
            .into();
    }
    if supervisor_alive(run) {
        return "running".into();
    }
    if run.join("pid").is_file() {
        return "crashed".into();
    }
    if epoch()
        - json(run.join("meta.json"))["startedEpoch"]
            .as_f64()
            .unwrap_or(0.0)
        <= 15.0
    {
        "starting".into()
    } else {
        "crashed".into()
    }
}
pub fn active(s: &str) -> bool {
    s == "running" || s == "starting"
}
pub fn events(run: &Path) -> Vec<Value> {
    read(run.join("events.jsonl"))
        .lines()
        .filter_map(|l| serde_json::from_str(l).ok())
        .collect()
}
pub fn next_step(run: &Path, meta: &Value, state: &str, sum: &Value) -> Option<String> {
    let name = if !s(meta, "run").is_empty() {
        s(meta, "run").to_string()
    } else {
        run.file_name()?.to_string_lossy().into_owned()
    };
    let script = script();
    let script = script.display();
    if active(state) {
        return Some(format!("{script} wait {name}"));
    }
    if state == "delivered" || state == "answered" {
        if s(meta, "mode") != "write" || n(&sum["changes"], "files") == 0 {
            return None;
        }
        if meta["worktree"].is_null() {
            return Some(format!(
                "review {script} diff {name}; the changes are already in the working tree"
            ));
        }
        if !run.join(".applied").exists() {
            return Some(format!(
                "review {script} diff {name} --total, then merge with {script} apply {name}"
            ));
        }
        return None;
    }
    match state {
        "rejected" => Some(format!(
            "read accept.tail; {script} reply {name} '<what to fix>' or take it over"
        )),
        "malformed" => Some("hand it to the other agent or do it yourself".into()),
        "failed" => Some("read error; fix the cause or take it over".into()),
        "timeout" => Some("split the task smaller or take it over".into()),
        "killed" | "crashed" => Some("take it over or start it again".into()),
        _ => None,
    }
}
pub fn status(run: &Path) -> Value {
    let meta = json(run.join("meta.json"));
    let st = state(run);
    let sum = json(run.join("summary.json"));
    let mut out = json!({"run":meta.get("run").and_then(Value::as_str).unwrap_or_else(||run.file_name().and_then(|x|x.to_str()).unwrap_or("")),"name":meta.get("name"),"state":st,"agent":meta.get("agent").and_then(Value::as_str).unwrap_or("pi"),"mode":meta.get("mode")});
    if !s(&meta, "parent").is_empty() {
        out["parent"] = meta["parent"].clone();
    }
    if meta["worktree"].is_object() {
        out["worktree"] = json!(s(&meta["worktree"], "path"));
    }
    if sum.is_object() {
        for key in [
            "elapsedSeconds",
            "attempts",
            "model",
            "turns",
            "files",
            "changes",
            "accept",
            "readOnlyViolation",
            "workspaceChanged",
            "queuedSeconds",
            "graceSeconds",
            "tokens",
            "warning",
            "error",
        ] {
            let v = &sum[key];
            if !v.is_null()
                && !v.as_array().is_some_and(Vec::is_empty)
                && !v.as_object().is_some_and(serde_json::Map::is_empty)
            {
                out[key] = v.clone();
            }
        }
    } else {
        let evs = events(run);
        let turns = evs.iter().filter(|x| s(x, "e") == "turn").count();
        out["elapsedSeconds"] =
            json!((epoch() - meta["startedEpoch"].as_f64().unwrap_or(epoch())).max(0.0) as i64);
        out["turns"] = json!(turns);
        let mut files = evs
            .iter()
            .filter(|x| ["edit", "write"].contains(&s(x, "e")) && !s(x, "path").is_empty())
            .map(|x| s(x, "path").to_string())
            .collect::<Vec<_>>();
        files.sort();
        files.dedup();
        if !files.is_empty() {
            out["files"] = json!(files);
        }
        if st == "running" {
            if let Some(last) = evs.iter().rev().find(|x| {
                !["turn", "result", "settled", "agent_end", "bash_done"].contains(&s(x, "e"))
            }) {
                let detail = ["cmd", "path", "arg", "detail"]
                    .iter()
                    .map(|k| s(last, k))
                    .find(|x| !x.is_empty())
                    .unwrap_or("");
                out["last"] = json!(format!(
                    "{}{}",
                    s(last, "e"),
                    if detail.is_empty() {
                        String::new()
                    } else {
                        format!(": {detail}")
                    }
                ));
                if let Some(at) = iso_epoch(s(last, "at")) {
                    out["idleSeconds"] = json!((epoch() as i64 - at).max(0));
                }
            }
        }
    }
    let result = run.join("result.md");
    if result.is_file() && fs::metadata(&result).is_ok_and(|m| m.len() > 0) {
        out["result"] = json!(result.to_string_lossy());
        out["resultChars"] = json!(read(&result).chars().count());
    }
    if let Some(step) = next_step(run, &meta, &st, &sum) {
        out["next"] = json!(step);
    }
    out["dir"] = json!(run.to_string_lossy());
    out
}
pub fn unmerged(run: &Path) -> bool {
    let m = json(run.join("meta.json"));
    s(&m, "mode") == "write"
        && m["worktree"].is_object()
        && Path::new(s(&m["worktree"], "path")).exists()
        && !run.join(".applied").exists()
}
pub fn remove(run: &Path) {
    let meta = json(run.join("meta.json"));
    let path = s(&meta["worktree"], "path").to_string();
    let _ = fs::remove_dir_all(run);
    if !path.is_empty()
        && !all_runs()
            .iter()
            .any(|r| s(&json(r.join("meta.json"))["worktree"], "path") == path)
    {
        worktree::remove(&meta);
    }
}
pub fn prune() {
    let Ok(days) = setting("KEEP_DAYS", "7").parse::<u64>() else {
        return;
    };
    if days == 0 {
        return;
    }
    for run in all_runs() {
        let done = run.join("exit_code");
        if run.join(".delivered").is_file() && done.is_file() && !unmerged(&run) {
            use std::time::UNIX_EPOCH;
            let age = fs::metadata(&done)
                .and_then(|x| x.modified())
                .ok()
                .and_then(|x| x.duration_since(UNIX_EPOCH).ok())
                .map(|x| epoch() - x.as_secs_f64())
                .unwrap_or(0.0);
            if age > days as f64 * 86400.0 {
                remove(&run);
            }
        }
    }
}
pub fn latest(mut run: PathBuf) -> PathBuf {
    loop {
        let mut replies = all_runs()
            .into_iter()
            .filter(|r| {
                s(&json(r.join("meta.json")), "parent")
                    == run.file_name().unwrap_or_default().to_string_lossy()
                    && state(r) != "malformed"
            })
            .collect::<Vec<_>>();
        if let Some(next) = replies.pop() {
            run = next;
        } else {
            return run;
        }
    }
}
